### PR TITLE
Replace ManualResetEvent with TaskCompletionSource in TestHost tests

### PR DESCRIPTION
**Pull request by Tratcher**
https://github.com/aspnet/AspNetCore-Internal/issues/1326#issuecomment-440348646
https://github.com/aspnet/AspNetCore-Internal/issues/1431

Addresses failures in ClientDisposalCloses and ClientCancellationAborts likely caused by deadlocks.

### CODE REVIEW

File: src/Hosting/TestHost/test/HttpContextBuilderTests.cs [hunk 1]
```diff
@@ -102,18 +102,17 @@ public async Task BlockingMiddlewareShouldNotBlockClient()
...
             var builder = new WebHostBuilder().Configure(app =>
             {
-                app.Run(c =>
+                app.Run(async c =>
                 {
                     c.Response.Headers["TestHeader"] = "TestValue";
                     var bytes = Encoding.UTF8.GetBytes("BodyStarted" + Environment.NewLine);
                     c.Response.Body.Write(bytes, 0, bytes.Length);
-                    Assert.True(block.WaitOne(TimeSpan.FromSeconds(5)));
+                    await block.Task;
```
Comment (halter73) [inline src/Hosting/TestHost/test/HttpContextBuilderTests.cs]: Nit: This had a timeout before. Should it still be there?

File: src/Hosting/TestHost/test/HttpContextBuilderTests.cs [hunk 2]
```diff
@@ -122,21 +121,21 @@ public async Task HeadersAvailableBeforeSyncBodyFinished()
...
-            var block = new ManualResetEvent(false);
+            var block = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
             var builder = new WebHostBuilder().Configure(app =>
             {
                 app.Run(async c =>
                 {
                     c.Response.Headers["TestHeader"] = "TestValue";
                     await c.Response.WriteAsync("BodyStarted" + Environment.NewLine);
-                    Assert.True(block.WaitOne(TimeSpan.FromSeconds(5)));
+                    await block.Task;
```
Comment (halter73) [inline src/Hosting/TestHost/test/HttpContextBuilderTests.cs]: ditto

**Review by halter73 — APPROVED** [review]

**Comment by Tratcher** [comment]
@muratg, test fixes only.

**Review by muratg — APPROVED** [review]
Approved.

**Pull request merged**